### DISCUSSION TITLE
Accuracy of setting stepper current in UART mode

### DISCUSSION THREAD

**Discussion by gothcha** [Jyers/Marlin#1024] [General Discussion]
Following from the Jyers gets Linear #707 I modified my 4.2.7 board for full UART and all appeared to perform well. However I started to investigate the current setting using Marlin. After digging into the code and the data sheet I found that there are some issues which I believe results in the currents not being set as specified. 

Edit
 
After extensive tests measuring the temperatures of the drivers it appears that the current setting is even more complex than I thought. There is additional PWM scaling being implemented which affects the calculations.

However the conclusion is that even though the Rsense value in Marlin (0.11) is not the actual value used on the Creality boards the 800mA nominal driver currents that are quoted have also been calculated using the same Rsense. So the current settings do appear to correspond and in fact the current can be set higher than the standalone values.

**Comment by Reggi3** [comment]
Did you change up the rsense values in the configs and recompile the firmware for testing?  I had noticed my Y motor has been a fair bit hotter (I think it's because I'm printing faster?) but dropping the current by 30ma seems to have resolved that issue for me.

I wonder if the marlin guys on discord have any insights, they have a dedicated trinamics channel...

**Reply by gothcha** [reply]
Yes I did, as I concluded the UART control is pretty complex and in StealthChop it uses an automatic tuning which is based on the CS value not on the Run current. I measured the temperature for the standalone mode then compared to that using UART with various Rsense and current values. I had read somewhere that the max current was defined by the trim pot but it turns out it isn’t. It does seem there is a wide scope for the current without issues, my Y axis settled around 47C. I ended up leaving the Rsense at 0.11 and setting the current for X&Z at 800 and Y at 730. 
Something I did find was that the E current setting in the adv.h did not work (other axis did), using M122 it always reports 800mA which I could not explain. The Rsense for E was being set. Best to check as E generally is set at a higher value. Using M906 E990 works correctly.

**Reply by Reggi3** [reply]
it's interesting what you say about E current setting in the adv.h, when I initially did the mod for the extruder and flashed the firmware, I used the default current setting 800ma, when I did the other 3 drivers, I had the correct current values for adv.h but E did actually come out at 800ma again, I thought it was a one off, set it via M906 E990 and forgot about it until you mentioned it.

I'll see what I can find out from the marlin people, we may get some insights into how to set the drivers up,  it's worth looking at how other boards with TMC uart drivers are configured too.

From the ender 3 adv.h for SKR mini E3 v2 using TMC2209 chips in uart mode:
#if AXIS_IS_TMC(X)
    #define X_CURRENT       580        // (mA) RMS current. Multiply by 1.414 for peak current.
    #define X_CURRENT_HOME  (X_CURRENT/2)  // (mA) RMS current for sensorless homing
    #define X_MICROSTEPS     16        // 0..256
    #define X_RSENSE          0.11
    #define X_CHAIN_POS      -1        // -1..0: Not chained. 1: MCU MOSI connected. 2: Next in chain, ...
    //#define X_INTERPOLATE  true      // Enable to override 'INTERPOLATE' for the X axis
  #endif

Same for Y/Z with E being set to 650

**Reply by gothcha** [reply]
Yes very interesting, the 580mA is much closer to the actual RMS current that the V2 motors are running at, I believe that the E3 does actually use 0.11R sense resistors, I found a schematic with the values on. 

A more precise setting for the V2 would be 580mA with Sense of 0.15, which gives a similar result to 800 with 0.11. I didn’t change mine because of the E current problem, 800 with 0.15 would overheat the extruder motor.  I will update the document to correct my errors and include the temperature measurements when I have done some more investigations. I too will try to track down the bug in E setting. If more users are going there be using UART mode it would be good to know that we are configuring it appropriately.

**Reply by gothcha** [reply]
@Reggi3 I have completed the testing and as expected setting the Rsense to the required 0.15 results in the correct currents being set. This is the final report. I have looked to find the E0 bug but it is so complex it is impossible!
[Stepper currents in UART mode.pdf](https://github.com/Jyers/Marlin/files/6680372/Stepper.currents.in.UART.mode.pdf)

**Reply by Reggi3** [reply]
Well done, that's fantastic, thank you!  Looks like I'll adding some information to my PDF too then :-)
